ace-link.el

```diff
@@ -413,41 +413,51 @@ If EXTERNAL is double prefix, browse in new buffer."
 (defun ace-link-gnus ()
   "Open a visible link in a `gnus-article-mode' buffer."
   (interactive)
-  (when (eq major-mode 'gnus-summary-mode)
-    (gnus-summary-widget-forward 1))
-  (let ((pt (avy-with ace-link-gnus
-              (avy-process
-               (ace-link--gnus-collect)
-               (avy--style-fn avy-style)))))
-    (ace-link--gnus-action pt)))
+  (save-selected-window
+    (when (eq major-mode 'gnus-summary-mode)
+      (if-let ((win (gnus-get-buffer-window gnus-article-buffer 'visible)))
+          (progn
+            (select-window win)
+            (select-frame-set-input-focus (window-frame win)))
+        (user-error "No article window found")))
+    (let ((pt (avy-with ace-link-gnus
+                (avy-process
+                 (ace-link--gnus-collect)
+                 (avy--style-fn avy-style)))))
+      (ace-link--gnus-action pt))))
 
 (defun ace-link--gnus-action (pt)
   (when (number-or-marker-p pt)
     (goto-char (1+ pt))
-    (widget-button-press (point))))
-
-(declare-function widget-forward "wid-edit")
-(declare-function gnus-summary-widget-forward "gnus-sum")
-(declare-function widget-button-press "wid-edit")
+    (cond ((< emacs-major-version 27)
+           (widget-button-press (point)))
+          ((and (eq mm-text-html-renderer 'shr)
```

Comment:
This could possibly already be required for Emacs 26.3 to open links in portions of the buffer rendered by shr.